opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/nodes/UaVariableTypeNode.java

```diff
@@ -217,4 +228,302 @@ public String getNodeVersion() {
   public void setNodeVersion(String nodeVersion) {
     setProperty(VariableTypeNodeProperties.NodeVersion, nodeVersion);
   }
+
+  /**
+   * @return a new {@link UaVariableTypeNodeBuilder}.
+   */
+  public static UaVariableTypeNodeBuilder builder(UaNodeContext context) {
+    return new UaVariableTypeNodeBuilder(context);
+  }
+
+  /**
+   * Build a {@link UaVariableTypeNode} using the {@link UaVariableTypeNodeBuilder} supplied to the
+   * {@code build} function.
+   *
+   * @param context the {@link UaNodeContext} to use.
+   * @param build a function that accepts a {@link UaVariableTypeNodeBuilder} and uses it to build
+   *     and return a {@link UaVariableTypeNode}.
+   * @return a {@link UaVariableTypeNode} built using the supplied {@link
+   *     UaVariableTypeNodeBuilder}.
+   */
+  public static UaVariableTypeNode build(
+      UaNodeContext context, Function<UaVariableTypeNodeBuilder, UaVariableTypeNode> build) {
+
+    UaVariableTypeNodeBuilder builder = new UaVariableTypeNodeBuilder(context);
+
+    return build.apply(builder);
+  }
+
+  public static class UaVariableTypeNodeBuilder implements Supplier<UaVariableTypeNode> {
+
+    private final List<AttributeFilter> attributeFilters = new ArrayList<>();
+
+    private final List<Reference> references = new ArrayList<>();
+
+    private NodeId nodeId;
+    private QualifiedName browseName;
+    private LocalizedText displayName;
+    private LocalizedText description = LocalizedText.NULL_VALUE;
+    private UInteger writeMask = UInteger.MIN;
+    private UInteger userWriteMask = UInteger.MIN;
+    private RolePermissionType[] rolePermissions;
+    private RolePermissionType[] userRolePermissions;
+    private AccessRestrictionType accessRestrictions;
+
+    private DataValue value;
+    private NodeId dataType;
+    private Integer valueRank;
+    private UInteger[] arrayDimensions;
+    private Boolean isAbstract;
+
+    private final UaNodeContext context;
+
+    public UaVariableTypeNodeBuilder(UaNodeContext context) {
+      this.context = context;
+    }
+
+    /**
+     * @see #build()
+     */
+    @Override
+    public UaVariableTypeNode get() {
+      return build();
+    }
+
+    /**
+     * Build and return the {@link UaVariableTypeNode}.
+     *
+     * <p>The following fields are required: NodeId, BrowseName, DisplayName.
+     *
+     * @return a {@link UaVariableTypeNode} built from the configuration of this builder.
+     */
+    public UaVariableTypeNode build() {
+      Preconditions.checkNotNull(nodeId, "NodeId cannot be null");
+      Preconditions.checkNotNull(browseName, "BrowseName cannot be null");
+      Preconditions.checkNotNull(displayName, "DisplayName cannot be null");
+
+      long hasTypeDefinitionCount =
+          references.stream()
+              .filter(r -> NodeIds.HasTypeDefinition.equals(r.getReferenceTypeId()))
+              .count();
+
+      if (hasTypeDefinitionCount == 0) {
```

Comment:
I am a bit confused regarding the settypedefinition() in builder. Type nodes don't have a type definition. Is it checking if an instance has multiple type definition? or maybe you want to check subtype reference to assure that the new type has only one parent?